### PR TITLE
feat: nightly auto-improve pipeline (sessions → GitHub issues → draft PRs)

### PR DESCRIPTION
**Pull request by dabowman**
## What

A nightly **auto-improve pipeline** that turns every Figmagent session into captured GitHub findings — and draft fix PRs — with no manual steps. This replaces the previous three-manual-agent flow (analyze → file issues → dispatch fixes).

```
A. extract-sessions --all-projects   every Figmagent session, every repo   → .claude/sessions-json/
B. /analyze-session  (looped)         efficiency/error audit                → improvement-tracker.md + .claude/plans/
C. sync-tracker-issues                tracker → GitHub issues (deduped)     → issues
D. /dispatch-fixes                    safe auto-fixable issues              → DRAFT PRs
```

Orchestrated by [`scripts/auto-improve.sh`](scripts/auto-improve.sh), triggered nightly by launchd ([`scripts/launchd/`](scripts/launchd/README.md)). Runs on demand via `bun run auto-improve`.

## Why it doesn't duplicate

The improvement tracker is the **deduplicated source of truth**, not the analysis docs. `sync-tracker-issues` maps each `### [ID]` entry to one issue, keyed on an embedded `/issues/N` ref first, then the `[ID]` title prefix. Re-runs are idempotent — active entries with no issue are created; `verified`/`resolved`/`implemented` entries are closed; tracker-vs-GitHub disagreements are reported as `DRIFT` (never auto-reopened without `--reopen`).

## New / changed

- **New:** `scripts/sync-tracker-issues.ts`, `scripts/refresh-manifest.ts` (shared manifest logic, replaces inline Python in the skill), `scripts/auto-improve.sh`, `scripts/launchd/*`, `.claude/commands/dispatch-fixes.md`
- **Changed:** `extract-sessions.ts` gains `--all-projects` (discovers sessions across all repos by detecting real `mcp__Figmagent__` tool calls); `analyze-session` skill, `CLAUDE.md`, `package.json`, `.gitignore`

## Safety

- Nothing is pushed to `main`; no ready-for-review PR is ever opened — Stage D produces **draft** PRs only, in isolated git worktrees, aborting on any lint/test/build failure (max 2/run).
- `lint`/`check`/`format` scoped to `src tests` so ephemeral worktrees under `.claude/worktrees/` don't trip Biome's nested-root scan.

## Verified live

The pipeline already ran end-to-end while authoring this:
- **Stage C** filed issues **#89–#99** (11 findings, correctly labeled, zero duplicates) and closed **#57**.
- **Drift reconciliation:** all 26 drift entries were audited against production code/docs (4 parallel verify agents, file:line evidence each). Every fix had shipped — the drift was stale tracker `Status` fields, not premature closes. All 26 marked `verified`; **0 reopened**. `sync-tracker-issues --dry-run` now reports **0 drift / 61 in-sync**.

## To activate the nightly job (not done in this PR)

```bash
cp scripts/launchd/com.figmagent.auto-improve.plist ~/Library/LaunchAgents/
launchctl bootstrap gui/$(id -u) ~/Library/LaunchAgents/com.figmagent.auto-improve.plist
```

## Test plan

- `bun run lint` ✅ · `bun run test` ✅ (342 pass) · `bun run build:plugin` ✅
- `bun run sync-issues --dry-run` → 0 drift / 61 in-sync
- `bun run extract-sessions --all-projects --list` → discovers live Figmagent sessions across all repos

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: scripts/sync-tracker-issues.ts [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+const commit = (): void => {
+	if (!curId) return;
+	const body = bodyLines.join("\n").trim();
+	// An entry is "resolved" (→ its GitHub issue should be closed) when it sits
+	// under Resolved, or its Status is verified / resolved / implemented. Note
+	// "partially implemented" and "mixed" deliberately stay open (work remains).
+	const resolved =
+		section === "resolved" || /^(verified|resolved|implemented)\b/i.test(curStatus);
+	const cleanTitle = curTitleLine.replace(/\s*—\s*\[(?:#|PR\b).*$/u, "").trim();
+	const issueRef = parseIssueRef(`${curTitleLine}\n${body}`);
```
Comment (dabowman) [inline scripts/sync-tracker-issues.ts]: **🔴 Correctness — closes the wrong GitHub issue (already fired on #57).**

`parseIssueRef` (L66) takes the *first* `/issues/N` found anywhere in `${curTitleLine}\n${body}`, so an entry whose header has no issue link but whose body cross-references another issue binds to that foreign number.

Live case: **`[TOOL-009]`** (`Status: implemented` → `resolved=true`, no `/issues` link in its header) has a *Follow-up* line linking `#57` — a separate, still-open issue about scan-path pagination. `parseIssueRef`→57, `resolveNum`→57, `wantOpen=false`, #57 open ⇒ `gh issue close 57` with a misleading *"Resolved in tracker (status: implemented)"*. The PR description's own "closed #57" is this close — it closed unrelated open work.

Compounding: `resolveNum` (L211-212) puts the prose-scraped `issueRef` *ahead* of the reliable `[ID]`-title match (`numberByPrefix`), so even a correct prefix mapping is overridden by a stray prose link.

**Fix:** parse a dedicated structured field (e.g. `- **Issue**: #N`, like Status/Priority/Category already are at L149-154) instead of scraping prose; and make `numberByPrefix.get(t.id)` the primary key with `issueRef` only as fallback.

File: scripts/auto-improve.sh [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+echo "[$(ts)] Stage A — extract Figmagent sessions across all repos"
+"$BUN" scripts/extract-sessions.ts --all-projects --compact --no-thinking --include-agents
+"$BUN" scripts/refresh-manifest.ts
+
+# ---- Stage B: analyze each unanalyzed figma session ------------------------
+echo "[$(ts)] Stage B — analyze sessions (cap $MAX_ANALYZE)"
+for ((n = 1; n <= MAX_ANALYZE; n++)); do
+  NEEDS="$("$BUN" scripts/refresh-manifest.ts --count 2>/dev/null)"
+  echo "[$(ts)]   sessions needing analysis: ${NEEDS:-?}"
+  [[ "${NEEDS:-0}" =~ ^[0-9]+$ ]] || { echo "[$(ts)]   manifest count unreadable — stopping Stage B"; break; }
```
Comment (dabowman) [inline scripts/auto-improve.sh]: **🔴 Correctness — a crashed/empty manifest count does NOT stop Stage B.** Verified under `/bin/bash`.

When `refresh-manifest.ts --count` prints nothing (e.g. its unconditional `writeFile` at refresh-manifest.ts:107 throws *before* the `--count` branch, or any uncaught error — stderr is swallowed by `2>/dev/null`), `$NEEDS` is empty:
- L56 `[[ "${NEEDS:-0}" =~ ^[0-9]+$ ]]` tests the defaulted `"0"` → **passes**, so the *"manifest count unreadable — stopping"* break never fires.
- L57 `[ "$NEEDS" -eq 0 ]` runs on the *bare* empty value → bash errors `integer expression expected` (rc 2), so `&& break` is skipped.

Net: the loop falls through and fires `claude -p "/analyze-session"` up to `MAX_ANALYZE` (25) times against a broken manifest. Separately, *any extra stdout line* makes `^[0-9]+$` fail and aborts Stage B on iteration 1.

**Fix:** `NEEDS="$(... 2>/dev/null | tail -n1)"`; test the raw value (`[[ "$NEEDS" =~ ^[0-9]+$ ]] || break`, then `[ "$NEEDS" -eq 0 ] && break`); and make `--count` resilient so a manifest-write failure can't masquerade as `0`.

File: scripts/sync-tracker-issues.ts [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+		// resolved is sticky; keep the first issue ref we saw.
+		if (body.length > existing.body.length) {
+			existing.body = body;
+			existing.cleanTitle = cleanTitle;
+			existing.fullTitle = `[${curId}] ${cleanTitle}`;
+			if (curStatus) existing.status = curStatus;
+			if (curPriority) existing.priority = curPriority;
+			if (curCategory) existing.category = curCategory;
+		}
+		existing.resolved = existing.resolved || resolved;
```
Comment (dabowman) [inline scripts/sync-tracker-issues.ts]: **🟠 Correctness — sticky `resolved` + stale `status` can close in-progress work.**

When the same ID appears in both `## Active Issues` and `## Resolved Issues`, the merge makes `resolved` sticky (`existing.resolved || resolved`, L117) but only refreshes `status`/`priority`/`category` inside the *longer-body* branch (L109-116). If the Active entry (e.g. `Status: identified`, `resolved=false`) is seen first with the richer body and a short Resolved recap follows, the entry ends up `resolved=true` with `status="identified"`. Reconcile then closes the open issue with the self-contradicting comment *"Resolved in tracker (status: identified)"*.

Also order-dependent: a Resolved recap parsed *before* a re-activated Active entry permanently pins `resolved=true` (it never flips back).

**Fix:** derive `resolved` from the authoritative occurrence's section/status rather than OR-ing across duplicates, and refresh status whenever an occurrence is more recent/resolved — not only when its body is longer.

File: scripts/extract-sessions.ts [hunk 4]
```diff
@@ -172,6 +173,101 @@ async function discoverSessions(): Promise<SessionInfo[]> {
...
+		let dirEntries: string[];
+		try {
+			dirEntries = await readdir(dir);
+		} catch {
+			continue;
+		}
+		const jsonlFiles = dirEntries.filter((e) => e.endsWith(".jsonl"));
+		for (const file of jsonlFiles) {
+			const filePath = join(dir, file);
+			const raw = await readFile(filePath, "utf-8");
```
Comment (dabowman) [inline scripts/extract-sessions.ts]: **🟠 Robustness — one unreadable transcript aborts the entire nightly extraction.**

In `discoverFigmaSessionsAllProjects`, `const raw = await readFile(filePath, "utf-8")` (L205) has no try/catch, unlike the `readdir` guards around it. `--all-projects` walks **every** repo's `~/.claude/projects/*/`, so a single `.jsonl` that is mid-write, rotated, or permission-restricted rejects → `main().catch` (L691) → `process.exit(1)`. auto-improve.sh runs this with no `|| true` (Stage A, L48), so under `set -uo pipefail` extraction produces nothing and the pipeline proceeds to refresh-manifest on a stale/partial set — silently skipping new sessions in other repos.

**Fix:** wrap the per-file `readFile` (and the `stat` below it) in try/catch and `continue` on error, like the single-file path does.

File: scripts/refresh-manifest.ts [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+		continue;
+	}
+	const sid = data.sessionId;
+	if (!sid) continue;
+
+	const m = data.metadata || {};
+	const tools = m.uniqueTools || [];
+	const figmaTools = tools.filter((t) => t.includes("Figmagent"));
+	const tc = m.toolCallCount || 0;
+	const sourceModified = mtimeSeconds((await stat(path)).mtimeMs);
```
Comment (dabowman) [inline scripts/refresh-manifest.ts]: **🟠 Robustness — unguarded `stat` can crash the whole manifest refresh.**

`const sourceModified = mtimeSeconds((await stat(path)).mtimeMs)` (L75) sits *outside* the try/catch that guards the `readFile`/`JSON.parse` above (L63-67), and unlike the analysis-file `stat` at L94 which *is* guarded. A TOCTOU (file removed/rotated between `readdir` at L48 and here) throws an unhandled rejection → non-zero exit with empty stdout — which is exactly the empty-`$NEEDS` input that defeats the Stage B loop guard in auto-improve.sh (see that finding).

**Fix:** move L75 inside the per-file try/catch and `continue` on failure.

File: scripts/refresh-manifest.ts [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+	manifest.sessions[sid] = entry;
+}
+
+await writeFile(MANIFEST, JSON.stringify(manifest, null, 2));
+
+const figma = Object.entries(manifest.sessions).filter(
+	([, v]) => v.sessionType === "figma",
+);
+const needs = figma
+	.filter(([, v]) => !v.analysis || (v.sourceModified ?? 0) > (v.analyzedAt ?? 0))
```
Comment (dabowman) [inline scripts/refresh-manifest.ts]: **🟠 Correctness — resumed sessions get re-analyzed every cycle (tracker churn).**

`sourceModified` (L75) is the mtime of the *extracted* `.claude/sessions-json/<id>.json`, and the needs-filter flags re-analysis when `sourceModified > analyzedAt` (L113). But extract-sessions re-extracts resumed/continued sessions (its own header comment says so), rewriting that JSON with a *now* mtime > the old `analyzedAt`. The session re-enters `needs`, Stage B re-runs `/analyze-session` on it, and the skill re-appends to the tracker (dedup is best-effort substring matching). Bounded by `MAX_ANALYZE`, but it burns the nightly budget on rework and grows the tracker/issue set.

**Fix:** key the freshness check on the *original transcript* mtime (or an analyzed-content hash), not the re-extracted JSON's mtime.

File: .claude/commands/dispatch-fixes.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+Open **draft** pull requests for a small, safe batch of auto-fixable issues from
+the improvement tracker. This runs unattended. **Be conservative: skipping an
+issue is always better than opening a bad PR.**
+
+## Hard constraints (do not violate)
+
+- **Draft PRs only.** Never open a ready-for-review PR. Never merge. Never push to `main`.
+- **Max 2 issues per run.**
+- Act on an issue only if **all** of these hold:
```
Comment (dabowman) [inline .claude/commands/dispatch-fixes.md:21]: **🟠 Integration — Stage D's gate matches almost nothing the analyzer produces.**

The constraints require `Auto-fixable: yes` (#1) and `Status: identified|planned` (#4). In the live tracker only **2 of 74** entries carry an `Auto-fixable` line, and **0** have `Status: planned` — even though the analyze-session skill's Phase 6 claims to set `planned` after writing a plan. So the documented "open draft fix PRs" stage is effectively inert.

**Fix:** make the analyze-session skill *reliably* emit `- **Auto-fixable**: yes/no` and set `Status: planned` once a plan file exists, or realign the gate vocabulary so it intersects what the skill actually writes. (Keep these two docs in lockstep — CLAUDE.md already notes the manifest logic is shared.)

File: scripts/auto-improve.sh [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+  "$CLAUDE" -p "/analyze-session" $PERMS \
+    || echo "[$(ts)]   /analyze-session exited non-zero (continuing)"
+done
+
+# ---- commit analysis artifacts so Stage D branches from a clean tree --------
+if [ "$DO_COMMIT" = "1" ]; then
+  if ! git diff --quiet -- .claude/analysis .claude/plans 2>/dev/null \
+     || [ -n "$(git ls-files --others --exclude-standard .claude/analysis .claude/plans)" ]; then
+    echo "[$(ts)] committing analysis artifacts (local, no push)"
+    git add .claude/analysis .claude/plans
```
Comment (dabowman) [inline scripts/auto-improve.sh:77]: **🟠 Correctness — auto-commit lands on the current branch, not `main` (and Stage D can't see it).**

The comments (L33 *"commit … to main"*, and the header) say main, but the script only `cd "$REPO_DIR"` then `git add`/`commit`s onto whatever HEAD points at — currently `feat/auto-improve-pipeline`. So nightly `auto-improve: session analyses` commits accrue on whatever feature branch is checked out, mixing pipeline bookkeeping into unrelated work. Worse, Stage D creates its worktree from `origin/main` (dispatch-fixes.md L45), so tonight's locally-committed analyses/plans are **not** on the base tonight's fix PRs build from.

**Fix:** check out/verify `main` before committing (or commit to a dedicated branch), and base Stage D's worktree on the same ref the analyses were committed to.

File: scripts/sync-tracker-issues.ts [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+// ---- read current GitHub state ---------------------------------------------
+
+interface GhIssue {
+	number: number;
+	title: string;
+	state: string;
+}
+
+const listJson =
+	await $`gh issue list --repo ${REPO} --state all --limit 1000 --json number,title,state`
```
Comment (dabowman) [inline scripts/sync-tracker-issues.ts]: **🟡 Correctness (latent) — `--limit 1000` will eventually duplicate issues.**

`gh issue list … --limit 1000` (L179) caps the snapshot. The pipeline files issues nightly from a growing tracker (74 entries today). Once total issues exceed 1000, an existing `[ID]`-titled issue past the window is absent from `numberByPrefix`/`stateByNumber` → `resolveNum` returns undefined for an active entry → the create branch files a **second** issue for an ID that already has one, every night — breaking the "never duplicates" guarantee.

**Fix:** scope the listing to the sync's own label (`--label figmagent-improvement`) and/or paginate; the embedded `/issues/N` fallback already covers refs outside that set.

File: scripts/sync-tracker-issues.ts [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+
+const TRACKER = ".claude/analysis/improvement-tracker.md";
+const REPO = process.env.AUTO_IMPROVE_REPO || "dabowman/Figmagent";
+const LABEL = "figmagent-improvement";
+
+const dryRun = process.argv.includes("--dry-run");
+const reopen = process.argv.includes("--reopen");
+const limitArg = process.argv.find((a) => a.startsWith("--limit="));
+const createLimit = limitArg
+	? Number.parseInt(limitArg.split("=")[1] ?? "", 10)
```
Comment (dabowman) [inline scripts/sync-tracker-issues.ts]: **🟡 Correctness — a malformed `--limit` silently disables the cap.** Verified.

`Number.parseInt(limitArg.split("=")[1] ?? "", 10)` (L48) yields `NaN` for `--limit=` (empty) or `--limit=abc`. At L281 `created >= createLimit` becomes `created >= NaN` → always `false`, so the cap never trips and *every* eligible entry is filed — the opposite of the operator's intent to throttle a noisy first sync.

**Fix:** validate the parse and fail/clamp on `NaN` (`Number.isNaN(n)` → error or treat as unset), reserving `Infinity` for the genuinely-absent flag.

File: scripts/refresh-manifest.ts [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+			} catch {
+				// analysis file was deleted — treat as needing re-analysis
+			}
+		}
+	} else {
+		entry.sessionType = "dev";
+		entry.skip = true;
+	}
+
+	manifest.sessions[sid] = entry;
```
Comment (dabowman) [inline scripts/refresh-manifest.ts:162]: **🟡 Correctness (latent) — stale manifest entries are never pruned.**

The refresh only *writes* entries for files currently on disk (L104) and writes the whole object back (L107); a session whose `.json` was later deleted keeps its prior entry. If that orphan is `sessionType:"figma"` with no analysis, the needs-filter (L112-114) keeps counting it and `--count` reports `>0` forever — so the Stage B `while count>0` loop can never reach 0 and burns `MAX_ANALYZE` `claude` runs/night on a phantom session it can't analyze. (0 orphans on disk today, so latent.)

**Fix:** during the refresh, drop manifest keys that have no corresponding on-disk session file.

File: .claude/commands/dispatch-fixes.md [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+  worktree, comment on the issue that auto-fix failed and needs manual work, open **no** PR.
+
+## Steps
+
+1. **Pick candidates.** Read `.claude/analysis/improvement-tracker.md`. Collect entries
+   meeting constraints 1–4 above. For each, `ls .claude/plans/` to confirm a plan file
+   exists. Cap the list at 2 (lowest issue numbers / highest priority first).
+
+2. **Confirm the GitHub issue is open and unclaimed.** For each candidate ID:
+   - `gh issue list --repo dabowman/Figmagent --state open --search "[<ID>]" --json number,title`
```
Comment (dabowman) [inline .claude/commands/dispatch-fixes.md]: **🟡 Consistency — `AUTO_IMPROVE_REPO` is honored by Stage C but hardcoded in Stage D.**

sync-tracker-issues.ts reads `process.env.AUTO_IMPROVE_REPO || "dabowman/Figmagent"` (L41), and the README documents it as a knob. But this command hardcodes `dabowman/Figmagent` in three `gh` calls (L36, L39, L69). Setting `AUTO_IMPROVE_REPO=other` then sends Stage C's issues to repo *other* while Stage D queries issues and opens PRs against `dabowman/Figmagent` — the pipeline splits across two repos.

**Fix:** thread the repo (env/arg) into the dispatch-fixes `gh` invocations instead of hardcoding it.

File: package.json [hunk 13]
```diff
@@ -7,13 +7,16 @@
     "socket": "bun run src/socket.ts",
     "setup": "./scripts/setup.sh",
     "test": "bun test",
-    "lint": "bunx biome lint .",
-    "lint:fix": "bunx biome check --write .",
-    "format": "bunx biome format --write .",
-    "check": "bunx biome check .",
+    "lint": "bunx biome lint src tests",
```
Comment (dabowman) [inline package.json]: **🟡 Altitude — the lint-scope change leaves the new mutation scripts with zero static checking.**

Narrowing `lint`/`check`/`format` from `.` to `src tests` is *needed* (confirmed: `biome lint .` errors on the nested `biome.json` files under `.claude/worktrees/`). But biome's `files.includes` is already `src/** + tests/** + figma_plugin`, so `scripts/` was *never* linted — and this PR adds `scripts/sync-tracker-issues.ts` + `refresh-manifest.ts` and edits `extract-sessions.ts`, all of which shell out to `gh issue close/create` and `git`, now with **no** lint/format/type checks. The Stage D draft-PR gate runs the same narrowed lint, so it can't catch a regression in these scripts either.

**Fix (deeper):** add `scripts/**/*.ts` to biome `files.includes` so the new code is linted, and exclude `.claude/worktrees/` via biome's ignore so the broad target survives — rather than permanently shrinking the lint surface.

File: .claude/commands/dispatch-fixes.md [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+allowed-tools: Bash, Read, Edit, Write, Glob, Grep
+---
+
+# Dispatch Fix PRs (Stage D of auto-improve)
+
+Open **draft** pull requests for a small, safe batch of auto-fixable issues from
+the improvement tracker. This runs unattended. **Be conservative: skipping an
+issue is always better than opening a bad PR.**
+
+## Hard constraints (do not violate)
```
Comment (dabowman) [inline .claude/commands/dispatch-fixes.md:12]: **🟡 Altitude — irreversible git/gh automation is driven by a prose prompt, not a deterministic script.**

Stage D's hard constraints (draft-only, max 2, never push to `main`, abort-on-failure, worktree cleanup) are natural-language instructions to `claude -p --dangerously-skip-permissions`, run unattended at 03:00 — with no enforcing code. A prompt-adherence lapse, a model update, or an instruction injected inside a tracker entry / plan file the agent reads could yield a non-draft PR, a push to the wrong branch, or a leftover worktree. Contrast Stage C, which is deterministic TypeScript.

**Fix:** move the mechanical git/gh steps (worktree add, branch, commit, push, `gh pr create --draft`) into a script that takes an issue ID + plan path, leaving only the *judgement* (does this plan apply cleanly?) to the model.

**Comment by dabowman** [comment]
Thanks for the thorough review — all 10 findings were real. Addressed in 1af7251. Mapping each:

### 🔴 Correctness

- **`sync-tracker-issues.ts` — closed the wrong issue (#57).** You're exactly right: `parseIssueRef` scraped the *first* `/issues/N` from `header + body`, so TOOL-009's "Follow-up: #57" cross-ref bound it to #57. Fixed: refs now come from the **header line only** or a structured `- **Issue**: #N` field — never body prose — and `numberByPrefix.get(id)` is now the **primary** key with the ref as fallback. **#57 reopened.**
- **`auto-improve.sh` — broken-manifest count didn't stop Stage B.** Fixed: `NEEDS="$(… | tail -n1)"`, and the guard tests the **raw** value (`[[ "$NEEDS" =~ ^[0-9]+$ ]] || break`) so empty/non-numeric stops the loop instead of falling through to fire `analyze` 25×.

### 🟠 Robustness / correctness

- **Sticky `resolved` + stale status.** Reworked: `resolved` is derived from the **authoritative Active-section occurrence's** status (a stale Resolved recap no longer pins it true), and the close comment uses a `resolvedReason` that can't contradict the status.
- **`extract-sessions.ts` — one bad transcript aborts extraction.** Per-file `readFile`/`stat` now wrapped in try/catch → `continue`.
- **`refresh-manifest.ts` — unguarded `stat` crash.** Moved inside the per-file try/catch.
- **`refresh-manifest.ts` — resumed-session churn.** Freshness now keys on the **original transcript's** mtime (derived from `cwd`+id), not the re-extracted JSON's. Verified: a continued session re-analyzes once then settles.
- **`auto-improve.sh` — auto-commit on the wrong branch.** Commit + Stage D now gated on `BRANCH = main`; off-main runs leave artifacts uncommitted and skip Stage D.
- **Stage D gate matched nothing.** The `analyze-session` skill now *always* emits `- **Auto-fixable**: yes/no` and sets `Status: planned` after writing a plan; `dispatch-fixes.md` notes the lockstep. (Also: the skill now increments IDs past the max across **both** sections — see the bonus below.)

### 🟡 Latent

- **`--limit 1000` → eventual dup.** Replaced with `gh api --paginate --slurp` (filters PRs) — no window cap.
- **Malformed `--limit` disabled the cap.** Now validates: `NaN`/negative exits with an error; `Infinity` reserved for the absent flag.

### Bonus found while fixing this

The 🟠 sticky-resolved fix surfaced a real data bug: the analyzer had **reused `AGENT-019`** for a distinct session-35 finding (two issues, one ID). Renumbered it to `AGENT-020`, and the sync now **detects ID collisions** (warns on same-ID/different-title) so it can't silently mis-merge again.

### Not addressed (out of scope)

The failing `check` job is **pre-existing** — `main`'s last three CI runs are red on the same `tests/connection.test.ts` (a WebSocket integration test, flaky in CI; passes locally 10/10). Not introduced here; worth a separate fix.

**Comment by dabowman** [comment]
**Validation follow-up (3b0717f).** You did a live manual run — it validated the critical fixes end-to-end:
- **#57 stayed open** (0 closes); Stage C created exactly the 4 genuine new findings (#104–#107), no dups.
- **Branch guard worked**: on a feature branch, the commit and Stage D were both skipped ("not on main").
- Loop guard converged 8→0; cross-repo extraction clean.

But the run also showed **finding #6 was only half-fixed** — the mtime key still flagged sessions 27/33/34/35 as 'updated' false positives (a transcript 13 min newer than its analysis, identical 12 calls), forcing the analyze skill to hand-reconcile the manifest. Replaced mtime with a **content signature** (`toolCalls:figmaToolCalls`), maintained inside refresh-manifest (migrate / resync / keep-prior). Now a stable 0 across runs and immune to touch-without-change, while a real content delta still re-flags.

**Comment by dabowman** [comment]
Actioned the four remaining 🟡 review items (plus the red CI) in c115388:

- **#11 — stale manifest entries never pruned.** `refresh-manifest.ts` now tracks session IDs present on disk and deletes any manifest key whose extracted JSON is gone, so an orphaned figma entry can no longer pin the Stage B needs-count above 0. Verified: a synthetic orphan is removed on the next refresh; 0 orphans today so the committed manifest is unchanged.
- **#12 + #14 — Stage D was prose-driven and hardcoded the repo.** Extracted the mechanical git/gh steps into a deterministic `scripts/dispatch-fix.ts` (`preflight` / `setup` / `publish` / `abort`). It hard-enforces *draft-only*, *base `main`*, *never push to `main`*, *worktree cleanup*, and the target repo via `AUTO_IMPROVE_REPO` (default `dabowman/Figmagent`, never hardcoded). `dispatch-fixes.md` now supplies only the judgement (pick the issue, does the plan apply); the irreversible mechanics live in code.
- **#13 — new scripts were unlinted.** Added `scripts/**/*.ts` to the Biome `files.includes` (and excluded `.claude/worktrees/**`), and pointed `lint`/`check`/`format` at `src tests scripts`. Fixed the warnings this surfaced (an unused `fileStat`, two `useOptionalChain`).

**CI:** the failing `check` job was the pre-existing flaky `tests/connection.test.ts` (same 3 `sendCommandToFigma` tests fail identically on `main` @ `33bee21f`, which predates this PR; they pass 10/10 locally). Root cause was fixed `setTimeout` waits (500ms relay spawn, 300ms WS connect) under-provisioning on loaded CI runners. Replaced with readiness polling — poll `/channels` until the relay answers, and a new exported `waitForConnection()` that awaits `ws` reaching OPEN. Suite is now 342/342 and ~2x faster locally.

**Comment by dabowman** [comment]
**CI fix — final approach (green as of b7eddb6).** The first de-flake attempt (readiness polling alone) actually made CI *worse* (5 fails) because the shared singleton connection let `connectToFigma` reuse a prior test's socket, so polling returned instantly and removed the accidental settle time that was masking the real bug. The real root cause was **shared state**, not timing:

1. The `sendCommandToFigma` tests share the singleton connection module *and* assert on relay channel state, so a channel left joined by another test (or a reused socket) made them order/load dependent — the no-channel assertion in particular auto-joined a leftover channel instead of rejecting.
2. Fix: each test now gets a **pristine relay on its own port** (zero channels) via `beforeEach`/`afterEach`, a freshly-disconnected module, and a real `waitForConnection()` readiness wait.
3. That surfaced an unhandled `ErrorEvent` on relay teardown (tests passed, process exited 1) — fixed by no-op `onerror` on the spy sockets and a no-op error handler kept through `disconnectFromFigma()` teardown (a socket mid-connect/close can still emit after its listeners are stripped; bun `latest` crashes on it, local 1.3.10 tolerated it).

`check` is now ✅ and all 14 review threads are resolved.
